[CUTLASS] Raise NotImplementedError if X & W aren't FixedLayout

ColinPeppler opened the pull request:
Summary: title

Differential Revision: D66131402




cc @voznesenskym @penguinwu @EikanWang @jgong5 @Guobing-Chen @XiaobingSuper @zhuhaozhe @blzheng @wenzhe-nrv @jiayisunx @ipiszy @yf225 @chenyang78 @kadeng @muchulee8 @amjames @desertfire @chauhang @aakhundov

aakhundov commented on torch/_inductor/codegen/cuda/gemm_template.py:
```diff
@@ -893,8 +893,11 @@ def render(  # type: ignore[override]
+            assert isinstance(X.layout, FixedLayout), "X.layout is not fixed"
+            assert isinstance(W.layout, FixedLayout), "W.layout is not fixed"
+        except AssertionError:
+            raise NotImplementedError
```
This seems less informative than the assertion error above? Is there a reason we need to raise `NotImplementedError` specifically?

ColinPeppler replied:
Agreed, I will try SkyLion's approach.

And good q, because `maybe_append_choice` requires `NotImplementedError` when choosing not to add a choice.

https://github.com/pytorch/pytorch/blob/34e420519df9c711ac997888470022cd7c4a90f6/torch/_inductor/codegen/common.py#L2687-L2700

ColinPeppler commented on torch/_inductor/codegen/cuda/gemm_template.py:
```diff
@@ -893,8 +893,11 @@ def render(  # type: ignore[override]
         X, W = self.input_nodes[0], self.input_nodes[1]
-        assert isinstance(X.layout, FixedLayout), "X.layout is not fixed"
-        assert isinstance(W.layout, FixedLayout), "W.layout is not fixed"
+        try:
+            assert isinstance(X.layout, FixedLayout), "X.layout is not fixed"
+            assert isinstance(W.layout, FixedLayout), "W.layout is not fixed"
+        except AssertionError:
+            raise NotImplementedError
```
Thanks for the idea. That's a better way of doing this.

ColinPeppler committed:
[CUTLASS] Raise NotImplementedError if X & W aren't FixedLayout

Summary: If we encounter a non-FixedLayout, this assertion will crash inductor. So, let's handle it gracefully.

Differential Revision: D66131402
